### DISCUSSION TITLE
Memory leak using EVP_* in multi-threaded environment.

### DISCUSSION THREAD

**Discussion by dcb1001** [openssl/openssl#25649] [Q&A]
I am trying the implement a HMAC class via openssl 3.3.2 using EVP_* apis. When I create an object of HMAC class and perform operations on this object in a seperate thread. I see a memory leak.
```
==956341==ERROR:` LeakSanitizer: detected memory leaks

Direct leak of 240 byte(s) in 1 object(s) allocated from:
    #0 0x563524944b69 in malloc /usr/local/src/conda/compiler-rt-packages-18.1.8/compiler-rt/lib/asan/asan_malloc_linux.cpp:68:3
    #1 0x7f4eeafa7fa1 in CRYPTO_malloc (/n/local/pchauhan/.micromamba/envs/36BDA407D6374A44D19415DD842687CB63AE2741652F0B784F628A313611649A/lib/libcrypto.so.3+0x220fa1)

SUMMARY: AddressSanitizer: 240 byte(s) leaked in 1 allocation(s).
<end of output>
Test time =   0.18 sec
```

HMAC constructor
```
OSSL_PARAM params[2];
params[0]     = OSSL_PARAM_construct_utf8_string(); // called with appropiate arguments.
params[1]     = OSSL_PARAM_construct_end();
m_mac         = EVP_MAC_fetch(NULL, "HMAC", NULL);
m_hmacContext = EVP_MAC_CTX_new(m_mac);
EVP_MAC_init(m_hmacContext, (const unsigned char*)key, keyLen, params); // key and keyLen are constructor argument.
```

HMAC destructor
```
EVP_MAC_CTX_free(m_hmacContext);
EVP_MAC_free(m_mac);
```

Valgring output
```
==837877== 240 bytes in 1 blocks are definitely lost in loss record 3 of 3
==837877==    at 0x4C29F73: malloc (vg_replace_malloc.c:309)
==837877==    by 0x5057FA1: CRYPTO_malloc (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x50580D9: CRYPTO_zalloc (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x5069F49: ossl_rcu_read_lock (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x4F8BCD2: CONF_modules_load (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x4F8C603: CONF_modules_load_file_ex (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x4F8C9B5: ossl_config_int (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x5056F1B: ossl_init_config_ossl_ (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x5E0520A: __pthread_once_slow (in /usr/lib64/libpthread-2.17.so)
==837877==    by 0x5069D59: CRYPTO_THREAD_run_once (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x50577DA: OPENSSL_init_crypto (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
==837877==    by 0x50683FC: ossl_provider_doall_activated (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)

```
What am  I missing here? which is leading to memory leak?

**Comment by mattcaswell** [comment]
This doesn't look like a "real" memory leak. This is memory allocated during libcrypto initialisation, i.e. its allocated in a call to `OPENSSL_init_crypto`:

````
==837877==    by 0x50577DA: OPENSSL_init_crypto (in /n/local/pchauhan/.micromamba/envs/1BA0305E6E190429F150E483C4119CF55E088592988D999E2BE3B7818C806780/lib/libcrypto.so.3)
````

This memory sticks around during the lifetime of libcrypto. Ordinarily we expect this to be cleaned up by the `OPENSSL_cleanup` function. This should normally be called automatically when the process exits. For some reason this doesn't seem to be happening for you.

Did you compile openssl yourself? If so how did you build it (what configuration options did you use)?

**Reply by dcb1001** [reply]
When I am using the HMAC in single threaded environment no leak happens which is because of the ```OPENSSL_cleanup``` which is invoked at the termination of the process invoked by the ```atExit``` handler.
But, when i create a pthread and do the computation on HMAC object in the thread, i get the above leak error. However, when i make a call to ```OPENSSL_thread_stop``` in the destructor of the HMAC object. There is no leakage. So, my assumption is the memory associated with the handlers isn't freed when the thread exits.
Also, i saw one of the openssl test is calling the ```OPENSSL_thread_stop``` API before returning from the thread.
```
static void *thread_run(void *arg)
{
    void (*f)(void);

    *(void **) (&f) = arg;

    f();
    OPENSSL_thread_stop();
 
   return NULL;
}

```
Dir: -```openssl/test/threadstest.h```

**Reply by mattcaswell** [reply]
Ah! Right. Threads can consume OpenSSL resources. Often this will be automatically cleaned up when the thread exits. But there are certain cases where this doesn't happen - in which case an explicit call to `OPENSSL_thread_stop` is required prior to thread exit.

See the docs on this page:

https://docs.openssl.org/3.4/man3/OPENSSL_init_crypto/

In particular the NOTES section talks about some scenarios where the "automatic" clean up will not work and explicit calls to `OPENSSL_thread_stop` are required.

**Reply by dcb1001** [reply]
Got it!
Is there a difference in calling the ```OPENSSL_thread_stop``` in the destructor of HMAC object or at the end of the thread? Also it seems the destruction of non-default library context is destroyed differently. For example: - thread handlers aren't removed in the non-default constructor
